### DISCUSSION TITLE
Why can't links have required properties?

### DISCUSSION THREAD

**Discussion by CodesInChaos** [geldata/gel#1727] [General]
What's the technical reason for that limitation?

Attempting to work around the limitation with `constraint expression on (exists(__subject__@index));` fails with "I
nternalServerError: cannot represent operator EXISTS in Postgres". (I didn't expect it to work, since if this worked you probably wouldn't have this limitation in the first place)

**Comment by elprans** [comment]
Support for required link properties was removed some time ago because of the difficulties in implementing them reasonably at the time (both internally and semantically).  We might consider restoring support.

> cannot represent operator EXISTS in Postgres

That's actually just missing support for `EXISTS` in simple expression transformer.  Should be easy to fix.

**Comment by adriangb** [comment]
I'd be +1 on this feature being implemented.

**Comment by rostero1** [comment]
Is the `constraint expression` the recommended workaround?

**Reply by rostero1** [reply]
Example:

```
module default {
  type Address {
    name: str;
  }

  abstract link coordinates {
    coordinates: tuple<lng: float32, lat: float32>;
  }

  type Place {
    address: Address {
      extending coordinates;
      constraint expression on (exists(__subject__@coordinates)) {
        errmessage := '{__subject__} requires coordinates.';
      }
    }
  }
}
```

**Reply by elprans** [reply]
Yes, that's right.  In fact, I think we should reconsider allowing linkprops to be `required`.  Opened #5571 to track.

**Reply by aqc-carlodri** [reply]
This workaround seems not to work for me, it gives this error:
```
error: cannot use aggregate functions or operators in a non-aggregating constraint
```
do you have any other suggestion to obtain a required link property?
